contracts/modules/staking/StakingPool.sol

```diff
@@ -463,86 +463,96 @@ contract StakingPool is IStakingPool, Multicall {
     emit StakeDeposited(msg.sender, amount, trancheId, tokenId);
   }
 
+  /// @notice Withdraws stake and rewards for a given token and specified tranches.
+  /// @dev The function processes the withdrawal of both stake and rewards for a given staking NFT (`tokenId`).
+  ///      Call StakingPoolViewer.getTokens to retrieve the relevant tranche IDs for a tokenId.
+  ///      A stake can only be withdrawn if the the associated tranche where it was deposited has expired
+  ///      Operates only when the contract is not paused.
+  /// @param tokenId The ID of the staking NFT representing the deposited stake and its associated rewards.
+  /// @param withdrawStake Whether to withdraw the total stake associated with the `tokenId`.
+  /// @param withdrawRewards Whether to withdraw the total rewards associated with the `tokenId`.
+  /// @param trancheIds An array of tranche IDs associated with the `tokenId`, used to specify which tranches to withdraw from.
+  /// @return withdrawnStake The total stake withdrawn across all specified tranche IDs for the given `tokenId`.
+  /// @return withdrawnRewards The total rewards withdrawn across all specified tranche IDs for the given `tokenId`.
   function withdraw(
     uint tokenId,
     bool withdrawStake,
     bool withdrawRewards,
     uint[] memory trancheIds
   ) public whenNotPaused returns (uint withdrawnStake, uint withdrawnRewards) {
 
-    uint managerLockedInGovernanceUntil = nxm.isLockedForMV(manager());
-
     // pass false as it does not modify the share supply nor the reward per second
     processExpirations(true);
 
-    uint _accNxmPerRewardsShare = accNxmPerRewardsShare;
-    uint _firstActiveTrancheId = block.timestamp / TRANCHE_DURATION;
+    WithdrawTrancheContext memory trancheContext;
+    trancheContext.withdrawStake = withdrawStake;
+    trancheContext.withdrawRewards = withdrawRewards;
+    trancheContext._accNxmPerRewardsShare = accNxmPerRewardsShare;
+    trancheContext._firstActiveTrancheId = block.timestamp / TRANCHE_DURATION;
+    trancheContext.managerLockedInGovernanceUntil = nxm.isLockedForMV(manager());
+    trancheContext.destination = tokenId == 0 ? manager() : stakingNFT.ownerOf(tokenId);
```

Comment:
There is quite a bit of refactoring here in `withdraw` because of stack too deep error. But the change is very simple:

previous:
```solidity
emit Withdraw(msg.sender, tokenId, trancheId, trancheStakeToWithdraw, trancheRewardsToWithdraw); // previous
```
current:
```solidity
emit Withdraw(trancheContext.destination, tokenId, trancheId, trancheStakeToWithdraw, trancheRewardsToWithdraw); // current change
```